### PR TITLE
Convert service account token controller to use a work queue

### PR DESCRIPTION
**Pull request by liggitt**
Converts the service account token controller to use a work queue. This allows parallelization of token generation (useful when there are several simultaneous namespaces or service accounts being created). It also lets us requeue failures to be retried sooned than the next sync period (which can be very long).

Fixes an issue seen when a namespace is created with secrets quotaed, and the token controller tries to create a token secret prior to the quota status having been initialized. In that case, the secret is rejected at admission, and the token controller wasn't retrying until the resync period.

### CODE REVIEW

**Comment by liggitt** [comment]
cc @deads2k

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 1]
```diff
@@ -103,206 +113,269 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
 }
 
-// serviceAccountDeleted reacts to a ServiceAccount deletion by deleting all corresponding ServiceAccountToken Secrets
-func (e *TokensController) serviceAccountDeleted(obj interface{}) {
-	serviceAccount, ok := obj.(*api.ServiceAccount)
-	if !ok {
-		// Unknown type. If we missed a ServiceAccount deletion, the
-		// corresponding secrets will be cleaned up during the Secret re-list
-		return
+func (e *TokensController) requeue(queue *workqueue.Type, key interface{}) {
```
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: @deads2k is there a cleaner way to do this? this also doesn't protect against a hotloop if you get more than N persistent failures

**Comment by liggitt** [comment]
cc @lavalamp for controller review (or nominate alternate)

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 1 above)
Comment (kargakis) [inline pkg/controller/serviceaccount/tokens_controller.go]: https://github.com/kubernetes/kubernetes/pull/23444

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 2]
```diff
@@ -332,7 +397,8 @@ func (e *TokensController) createSecret(serviceAccount *api.ServiceAccount) erro
 
 	// Save the secret
 	if createdToken, err := e.client.Core().Secrets(serviceAccount.Namespace).Create(secret); err != nil {
-		return err
+		// retriable error
+		return true, err
 	} else {
```
Comment (kargakis) [inline pkg/controller/serviceaccount/tokens_controller.go]: Reduce indent

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 1 above)
Comment (deads2k) [inline pkg/controller/serviceaccount/tokens_controller.go]: > #23444

That in combination with a `ratelimit.NewBucketWithRate()` and then `bucket.Take()` is what I planned to do downstream.  I think that should give the duration to add to the requeue.  MaxFailures should be tracked both on your object (which may fail) and a list of those failures in your controller (which should not fail).

File: cmd/kube-controller-manager/app/options/options.go [hunk 3]
```diff
@@ -53,6 +53,7 @@ func NewCMServer() *CMServer {
 			ConcurrentResourceQuotaSyncs:      5,
 			ConcurrentDeploymentSyncs:         5,
 			ConcurrentNamespaceSyncs:          2,
+			ConcurrentSATokenSyncs:            5,
```
Comment (lavalamp) [inline cmd/kube-controller-manager/app/options/options.go]: "Workers" instead of Syncs?

Comment (lavalamp) [inline cmd/kube-controller-manager/app/options/options.go]: I guess that would be inconsistent and I won't ask you to change them all. Sigh. "Syncs" is confusing.

Comment (liggitt) [inline cmd/kube-controller-manager/app/options/options.go]: agree, and agree

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 4]
```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
 	serviceAccountController *framework.Controller
 	secretController         *framework.Controller
 
-	// These are here so tests can inject a 'return true'.
-	serviceAccountsSynced func() bool
-	secretsSynced         func() bool
+	// syncServiceAccountQueue handles service account events:
+	//   * ensures a referenced token exists for service accounts which still exist
+	//   * ensures tokens are removed for service accounts which no longer exist
+	// key is "<namespace>/<name>/<uid>"
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Why not just UID?

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 5]
```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
+	syncServiceAccountQueue *workqueue.Type
+
+	// syncSecretQueue handles secret events:
+	//   * deletes tokens whose service account no longer exists
+	//   * updates tokens with missing token or namespace data, or mismatched ca data
+	//   * ensures service account secret references are removed for tokens which are deleted
+	// key is "<namespace>/<name>/<uid>/<serviceaccount-name>/<serviceaccount-uid>" (serviceaccount-uid may be blank)
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Is it OK if `ns/n/uid/account-name/uid` and `ns/n/uid/account-name/` are processed concurrently?

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 4 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: because I have to look it up from the store by namespace/name

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 6]
```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
 	}
 }
 
-// secretAdded reacts to a Secret create by ensuring the referenced ServiceAccount exists, and by adding a token to the secret if needed
-func (e *TokensController) secretAdded(obj interface{}) {
-	secret := obj.(*api.Secret)
-	serviceAccount, err := e.getServiceAccount(secret, true)
-	if err != nil {
-		glog.Error(err)
+func (e *TokensController) syncServiceAccount() {
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: syncOneServiceAccount or serviceAccountWorker?

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 7]
```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
 	}
-	if newServiceAccount == nil {
-		glog.V(2).Infof(
-			"Deleting updated secret %s/%s because service account %s (uid=%s) was not found",
-			newSecret.Namespace, newSecret.Name,
-			newSecret.Annotations[api.ServiceAccountNameKey], newSecret.Annotations[api.ServiceAccountUIDKey])
-		if err := e.deleteSecret(newSecret); err != nil {
-			glog.Errorf("Error deleting secret %s/%s: %v", newSecret.Namespace, newSecret.Name, err)
+
+	sa, err := e.getServiceAccount(ns, name, uid, false)
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Ah, this is why you put all that stuff in the key.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 8]
```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
+	switch {
+	case err != nil:
+		// TODO: requeue?
+		glog.Error(err)
+	case sa == nil:
+		// service account no longer exists, so delete related tokens
+		glog.V(4).Infof("syncServiceAccount(%s/%s), service account deleted, removing tokens", ns, name)
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: How is this behavior triggered on controller startup (i.e., if you crash after the account is deleted but before this code manages to run?)

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 7 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: yeah, not sure how else to sync a deletion

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 8 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: https://github.com/kubernetes/kubernetes/pull/23858/files/b8f98d9a14e121eb243af96611a9ecb304179bb9#diff-4f732a5805bff59709b3d59deb588688R291 cleans up secrets whose referenced service account no longer exists

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 6 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: one

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 9]
```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
+			retry = true
 		}
 	}
+	return retry, utilerrors.NewAggregate(errs)
+}
 
-	// Otherwise create a new token secret
-	return e.createSecret(serviceAccount)
+func (e *TokensController) deleteToken(ns, name string) ( /*retry*/ bool, error) {
+	err := e.client.Core().Secrets(ns).Delete(name, nil)
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Consider setting a UID precondition.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 10]
```diff
@@ -341,31 +407,27 @@ func (e *TokensController) createSecret(serviceAccount *api.ServiceAccount) erro
...
 	liveServiceAccount.Secrets = append(liveServiceAccount.Secrets, api.ObjectReference{Name: secret.Name})
 
-	_, err = serviceAccounts.Update(liveServiceAccount)
-	if err != nil {
+	if _, err = serviceAccounts.Update(liveServiceAccount); err != nil {
 		// we weren't able to use the token, try to clean it up.
-		glog.V(2).Infof("Deleting secret %s/%s because reference couldn't be added (%v)", secret.Namespace, secret.Name, err)
+		glog.V(2).Infof("deleting secret %s/%s because reference couldn't be added (%v)", secret.Namespace, secret.Name, err)
+
 		if err := e.client.Core().Secrets(secret.Namespace).Delete(secret.Name, nil); err != nil {
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: UID precondition?

Also, if you crash here, will this secret get cleaned up in the future? If so, maybe just have a comment stating that and let that mechanism clean up the secret? If not, that seems problematic?

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 9 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: that's nifty, I didn't know we had those

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 11]
```diff
@@ -435,59 +540,90 @@ func (e *TokensController) removeSecretReferenceIfNeeded(serviceAccount *api.Ser
...
+	return nil, nil
+}
 
-	if fetchOnCacheMiss {
-		serviceAccount, err := e.client.Core().ServiceAccounts(secret.Namespace).Get(name)
-		if apierrors.IsNotFound(err) {
-			return nil, nil
+func (e *TokensController) getSecret(ns string, name string, uid types.UID, fetchOnCacheMiss bool) (*api.Secret, error) {
+	// Look up in cache
+	obj, exists, err := e.secrets.GetByKey(ns + "/" + name)
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: hidden dependency on the metanamespacekey function. Maybe add a function to do this construction.

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 2 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: if you mean remove the `else`, I'd rather keep `createdToken` and `err` scoped to this block

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 10 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: It will get cleaned up if the service account goes away, but not otherwise. I'm not sure how concerned about this particular clean up I am

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 12]
```diff
@@ -520,3 +656,50 @@ func getSecretReferences(serviceAccount *api.ServiceAccount) sets.String {
 	}
 	return references
 }
+
+func makeServiceAccountKey(sa *api.ServiceAccount) string {
+	return fmt.Sprintf("%s/%s/%s", sa.Namespace, sa.Name, sa.UID)
+}
+
+func parseServiceAccountKey(key interface{}) ( /*ns*/ string /*name*/, string /*uid*/, types.UID, error) {
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: optional nit: I think actually naming the parameters would be fine? the other parse function, too.

Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: optional nit: I think actually naming the parameters would be fine? the other parse function, too.

Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: I'm actually leaning toward making a struct as my key instead of munging everything into a string and then parsing it back out

**Comment by lavalamp** [comment]
We may need a strategy for testing this sort of code. There's a lot of possible places for error returns, are they all exercised?

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 12 above)
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: That would be much better, of course.

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 9 above)
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: They are quite new :)

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 5 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: yes

**Comment by liggitt** [comment]
comments addressed, added rate limiting and max retry count

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 13]
```diff
@@ -103,206 +141,309 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
 	}
-	for _, secret := range secrets {
-		glog.V(4).Infof("Deleting secret %s/%s because service account %s was deleted", secret.Namespace, secret.Name, serviceAccount.Name)
-		if err := e.deleteSecret(secret); err != nil {
-			glog.Errorf("Error deleting secret %s/%s: %v", secret.Namespace, secret.Name, err)
-		}
+
+	// Or delay
+	glog.V(4).Infof("rate-limiter delay of %s retrying %#v", key, delay)
+	go func() {
```
Comment (deads2k) [inline pkg/controller/serviceaccount/tokens_controller.go]: I think the fanout here is excessive, especially in cases of environmental failures happening on initial activation.  I'd still say that https://github.com/kubernetes/kubernetes/pull/23444 is a more appropriate place to perform the requeue action and probably a better place to do the overall rate limiting too.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 14]
```diff
@@ -193,23 +225,50 @@ func (e *TokensController) queueServiceAccountUpdateSync(oldObj interface{}, new
...
+		queue.Add(key)
+		return
 	}
-	queue.Add(key)
+
+	// Or delay
+	glog.V(4).Infof("rate-limiter delay of %s retrying %#v", key, delay)
+	go func() {
+		defer utilruntime.HandleCrash()
+		time.Sleep(delay)
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: I'm pretty sure this won't play nicely with the shutdown path?

Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: @deads2k convinced me the fanout wasn't good... I'll rebase once https://github.com/kubernetes/kubernetes/pull/23444 is available and use that instead

**Comment by lavalamp** [comment]
This is up next in my queue, will probably get to it tomorrow. Sorry for the delay.

**Comment by liggitt** [comment]
np, I've been sidetracked as well

**Comment by deads2k** [comment]
> np, I've been sidetracked as well

Do you want to see if https://github.com/kubernetes/kubernetes/pull/24052 satisfies your need cleanly?  This won't be the only one.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 15]
```diff
@@ -57,19 +62,53 @@ type TokensControllerOptions struct {
...
-	// This CA will be added in the secretes of service accounts
+	// This CA will be added in the secrets of service accounts
 	RootCA []byte
+
+	// RetryQPS controls the rate at which failed syncs are retried
+	// If zero, a default QPS is used
+	RetryQPS float64
+	// RetryBurst controls the rate at which failed syncs are retried
+	// If zero, a default burst is used
+	RetryBurst int64
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Is there any good reason to have a burst here other than 1? Just trying to make the options simpler

**Comment by lavalamp** [comment]
OK, I'm sure I missed stuff because this is so large, but LGTM.

I think it may be good to test stuff like this by starting it and using something like golang.org/pkg/testing/quick/ to generate work for it to do.

I guess if this causes a problem our e2es will probably hit it in short order.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 16]
```diff
@@ -57,19 +62,53 @@ type TokensControllerOptions struct {
...
 	// If zero, re-list will be delayed as long as possible
 	SecretResync time.Duration
-	// This CA will be added in the secretes of service accounts
+	// This CA will be added in the secrets of service accounts
 	RootCA []byte
+
+	// RetryQPS controls the rate at which failed syncs are retried
+	// If zero, a default QPS is used
+	RetryQPS float64
+	// RetryBurst controls the rate at which failed syncs are retried
```
Comment (stevekuznetsov) [inline pkg/controller/serviceaccount/tokens_controller.go]: GoDoc copy-pasta

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 17]
```diff
@@ -103,206 +141,300 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
-	}
-
-	// We shouldn't try to validate secret references until the secrets store is synced
-	if !e.secretsSynced() {
-		return nil
-	}
-
-	// If any existing token secrets are referenced by the service account, return
-	allSecrets, err := e.listTokenSecrets(serviceAccount)
+func (e *TokensController) deleteTokens(serviceAccount *api.ServiceAccount) ( /*retry*/ bool, error) {
```
Comment (stevekuznetsov) [inline pkg/controller/serviceaccount/tokens_controller.go]: Seems weird to see this instead of a named return value.

Comment (deads2k) [inline pkg/controller/serviceaccount/tokens_controller.go]: > Seems weird to see this instead of a named return value.

Named return values make code that returns early more difficult to reason about, especially when you depend on the default values for those fields.  After the first few bugs, we started avoiding them.

Comment (kargakis) [inline pkg/controller/serviceaccount/tokens_controller.go]: Right, please don't use them.

Comment (stevekuznetsov) [inline pkg/controller/serviceaccount/tokens_controller.go]: Named returns don't preclude you from returning explicit values, but I can see that this would be easier to enforce on reviews.

Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Personally, I prefer named return values, but I'll accept it either way.

**Comment by liggitt** [comment]
rebased on ratelimiting workqueue, tagging per [comment](https://github.com/kubernetes/kubernetes/pull/23858#issuecomment-212651333), but leaving out of the milestone

**Comment by liggitt** [comment]
fixed govet error

**Comment by liggitt** [comment]
fixed mesos controllermanager int32 cast

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 18]
```diff
@@ -58,21 +61,36 @@ type TokensControllerOptions struct {
...
+		syncSecretQueue:         workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		maxRetries: maxRetries,
 	}
 	if cl != nil && cl.Core().GetRESTClient().GetRateLimiter() != nil {
 		metrics.RegisterMetricAndTrackRateLimiterUsage("serviceaccount_controller", cl.Core().GetRESTClient().GetRateLimiter())
 	}
-	e.serviceAccounts, e.serviceAccountController = framework.NewIndexerInformer(
+
+	e.serviceAccounts, e.serviceAccountController = framework.NewInformer(
```
Comment (kargakis) [inline pkg/controller/serviceaccount/tokens_controller.go]: Why this change?

Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: we don't need to look up service accounts by namespace, so no need to maintain that index

**Comment by liggitt** [comment]
rebased

**Comment by liggitt** [comment]
rebased

**Comment by ncdc** [comment]
@k8s-bot unit test this issue: #28076

**Comment by liggitt** [comment]
@k8s-bot unit test this issue: #28076
